### ISSUE TITLE
Create profile directory if it doesn't exist and `--profile-create-new` has been combined to the `-p` cli option

### ISSUE THREAD

**Issue by blimmer** [mozilla/web-ext#2058]
### Is this a feature request or a bug?
Feature request

### What is the current behavior?

Today, if I pass a profile that does not exist, the `web-ext run` command fails:

```console
> web-ext run --source-dir ./dist/ -p extension_development
Applying config file: ./package.json
Running web extension from /Users/a70e7cc5/code/proj/dist

o: Could not copy Firefox profile from extension_development: Error: cannot find profile extension_development
    at Module.C (/Users/a70e7cc5/code/proj/node_modules/web-ext/dist/web-ext.js:1:22412)
```

### What is the expected or desired behavior?

It would be nice if there was a flag to create the profile if it doesn't exist. Even better, it would be nice to be able to pass a directory to create the profile in.

I'm able to work around this in my launch script:

```ts
import webExt from 'web-ext';
import { default as fxRunner } from 'fx-runner';

async function launchFirefox(browserProfileDirectory: string) {
  await createFirefoxProfileIfNotExists(browserProfileDirectory);
  await webExt.cmd.run({
    sourceDir: BUILD_DIR,
    firefoxProfile: browserProfileDirectory,
    keepProfileChanges: true,
    startUrl: ['about:debugging#/runtime/this-firefox'],
  });
}

async function createFirefoxProfileIfNotExists(browserProfileDirectory: string): Promise<void> {
  const devProfileName = 'extension_development';

  if (!fs.existsSync(browserProfileDirectory)) {
    console.info(`${devProfileName} Firefox profile does not exist. Creating...`);

    // The `-CreateProfile` flag creates a new Firefox profile stub and immediately exits
    const createProfile = await fxRunner({
      'no-remote': true,
      'binary-args': ['-CreateProfile', `${devProfileName} ${browserProfileDirectory}`],
    });
    const createProfileProc = createProfile.process;
    return new Promise((resolve, reject) => {
      createProfileProc.on('close', resolve);
      createProfileProc.on('error', reject);
    });
  }
}
```

### Version information (for bug reports)

* **Firefox version**: 83.0b4
* **Your OS and version**: MacOS 10.15.7
* Paste the output of these commands:

````
> node --version && npm --version && web-ext --version
v12.18.4
6.14.6
5.3.0
````

**Comment by rpl** [comment]
Having this behavior as an optional opt-in feature, e.g. enabled by a new `--profile-create-new`, sounds reasonable to me too.

**Comment by rpl** [comment]
mentor: @rpl 

To fix this issue we want to:
- define a new command line option for the `web-ext run` command
- when the new `--profile-create-new` option is specified:
  - the value set on `--profile` is going to be considered as a directory path
  - if the directory already exit, exit immediately, show to the user an error message to warn the user
  - if the directory doesn't exit, create it and use it as the firefox profile for the extension runner
- cover the new expected behavior with new unit tests

We should also double-check the behavior of the similar `--chrome-profile` option, because it may be reasonable to also behave consistently when it is combined with the `--profile-create-new` cli option. 

Follows some of the web-ext source code that may be interested by this change:
- The web-ext command line options are all defined in src/program.js (https://github.com/mozilla/web-ext/blob/55931e7b082d07734c37b8748798db20cc45b079/src/program.js#L542)
- The `web-ext run` command is implemented in `src/cmd/run.js` (https://github.com/mozilla/web-ext/blob/55931e7b082d07734c37b8748798db20cc45b079/src/cmd/run.js#L68)
- Existing internal helper methods used to find and manage the Firefox profile are defined in `src/firefox/index.js`
- Existing tests for the src/cmd/run.js module are defined in `tests/unit/test-cmd/test.run.js`

**Comment by akhilpanchal** [comment]
Hi @rpl 
I am interested in working on this. This is my first time contributing to this project, so I might need some time to familiarize myself with the project. I shall ask if I have any questions. Any tips or guidance to start would be really helpful.

**Comment by rpl** [comment]
> Hi @rpl
> I am interested in working on this. This is my first time contributing to this project, so I might need some time to familiarize myself with the project. I shall ask if I have any questions or face any issues.

@akhilpanchal Sure thing!

Given that this would be your first contribution, I suggest you to also:
- take a look to the contribution guide here: https://github.com/mozilla/web-ext/blob/master/CONTRIBUTING.md
- try to run the existing tests before making any chance, to be sure that the development environment is worked correctly

After that you should make sure to be able to reproduce (and then understand) the issue described in this issue, and then look to the pointers provided in https://github.com/mozilla/web-ext/issues/2058. 

Feel free to ask for help if you have any specific questions or issues.

**Comment by ankushduacodes** [comment]
@rpl is it okay if I work on this issue as well?

**Comment by akhilpanchal** [comment]
@ankushduacodes I have already started working on this issue. I should have a PR soon.

**Comment by akhilpanchal** [comment]
@rpl 
I am able to reproduce the issue. 
However I am a little unsure if I understand the role of this new option.

Is the expectation to create a directory and also create a new Firefox Profile in this directory, so that this newly created profile persists in the directory after the `web-ext run` process exits?
OR
is the expectation to just create a directory if it the given directory doesn't exist so that web-ext run can proceed without complaining that the profile directory doesn't exist?

**Comment by blimmer** [comment]
For my use case, the latter would be preferable. The profile gets created if it doesn't already exist.

**Comment by rpl** [comment]
> @rpl
> I am able to reproduce the issue.
> However I am a little unsure if I understand the role of this new option.
> 
> Is the expectation to create a directory and also create a new Firefox Profile in this directory?
> OR
> is the expectation to just create a directory if it the given directory doesn't exist so that web-ext run can proceed without complaining that the profile directory doesn't exist?

@akhilpanchal
:+1: I think that (when the new option is being used) we should just have to create that directory if it doesn't exist already, everything else should work as expected without any additional special handling.

Let's start with doing that (and create a new test case to cover those changes), as part of the pull request review we will also double-check (while looking to the actual changes) if there is anything more that should be done.

**Comment by blimmer** [comment]
I saw this got released with 5.4.0 and just started using it. Thank you to everyone who helped out with this!
